controllers/runner_controller.go

```diff
@@ -630,7 +630,7 @@ func newRunnerPod(runnerName string, template corev1.Pod, runnerSpec v1alpha1.Ru
 	pod := template.DeepCopy()
 
 	if pod.Spec.RestartPolicy == "" {
-		pod.Spec.RestartPolicy = "OnFailure"
+		pod.Spec.RestartPolicy = corev1.RestartPolicyNever
 	}
```

Comment:
add a comment for why we need to use `RestartPolicyNever` instead of another restart policy

Reply:
Thanks for your suggestion! Addressed in c36e99a.

Also, I realized that we can make it work for RunnerSet too by forcing restartPolicy=Never in the mutating webhook. Updated code accordingly, so that the initial description of this pull request that stated this change works only for RunnerDeployment is no longer true.

Note that, as we can now force restartPolicy=Never on every runner pod, we may never see the code path added in https://github.com/actions-runner-controller/actions-runner-controller/pull/1420 is executed in real-world scenarios.